Mathlib/NumberTheory/LSeries/ZMod.lean

```diff
@@ -0,0 +1,204 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+
+import Mathlib.Analysis.Fourier.ZMod
+import Mathlib.Analysis.NormedSpace.Connected
+import Mathlib.LinearAlgebra.Dimension.DivisionRing
+import Mathlib.NumberTheory.LSeries.RiemannZeta
+import Mathlib.Topology.Algebra.Module.Cardinality
+
+/-!
+# L-series of functions on `ZMod N`
+
+We show that if `N` is a positive integer and `Φ : ZMod N → ℂ`, then the L-series of `Φ` has
+analytic continuation (away from a pole at `s = 1` if `∑ j, Φ j ≠ 0`).
+
+The most familiar case is when `Φ` is a Dirichlet character, but the results here are valid
+for general functions; for the specific case of Dirichlet characters see
+`Mathlib.NumberTheory.LSeries.DirichletContinuation`.
+
+## Main definitions
+
+* `ZMod.LFunction Φ s`: the meromorphic continuation of the function `∑ n : ℕ, Φ n * n ^ (-s)`.
+
+## Main theorems
+
+* `ZMod.LFunction_eq_LSeries`: if `1 < re s` then the `LFunction` coincides with the naive
+  `LSeries`.
+* `ZMod.differentiableAt_LFunction`: `ZMod.LFunction Φ` is differentiable at `s ∈ ℂ` if either
+  `s ≠ 1` or `∑ j, Φ j = 0`.
+* `ZMod.LFunction_one_sub`: the functional equation relating `LFunction Φ (1 - s)` to
+  `LFunction (𝓕 Φ) s`, where `𝓕` is the Fourier transform.
+-/
+
+open HurwitzZeta Complex ZMod Finset Classical Topology Filter
+
+open scoped Real
+
+namespace ZMod
+
+variable {N : ℕ} [NeZero N]
+
+/-- If `Φ` is a periodic function, then the L-series of `Φ` converges for `1 < re s`. -/
+lemma LSeriesSummable_of_one_lt_re (Φ : ZMod N → ℂ) {s : ℂ} (hs : 1 < re s) :
+    LSeriesSummable (Φ ·) s := by
+  let c := max' _ <| univ_nonempty.image (Complex.abs ∘ Φ)
+  refine LSeriesSummable_of_bounded_of_one_lt_re (fun n _ ↦ le_max' _ _ ?_) (m := c) hs
+  exact mem_image_of_mem _ (mem_univ _)
+
+/--
+The unique meromorphic function `ℂ → ℂ` which agrees with `∑' n : ℕ, Φ n / n ^ s` wherever the
+latter is convergent. This is constructed as a linear combination of Hurwitz zeta functions.
+
+Note that this is not the same as `LSeries Φ`: they agree in the convergence range, but
+`LSeries Φ s` is defined to be `0` if `re s ≤ 1`.
+ -/
+noncomputable def LFunction (Φ : ZMod N → ℂ) (s : ℂ) : ℂ :=
+  N ^ (-s) * ∑ j : ZMod N, Φ j * hurwitzZeta (toAddCircle j) s
+
+example {α : Type} [Fintype α] [Unique α] (a : α) : (univ : Finset α) = {a} := by
```

Comment:
I'm guessing this example shouldn't be here.

Reply:
Oops! Yes, that was just me using `exact?` to find a simp lemma. Fixed now.